core/controllers/cron_test.py

```diff
@@ -0,0 +1,360 @@
+        self.admin_id = self.get_user_id_from_email(self.ADMIN_EMAIL)
+        self.set_admins([self.ADMIN_USERNAME])
+        self.testapp_swap = self.swap(
+            self, 'testapp', webtest.TestApp(main_cron.app))
+
+    def test_send_mail_to_admin_on_job_success(self):
+        self.login(self.ADMIN_EMAIL, is_super_admin=True)
+
+        email_subjects = []
+        email_bodies = []
+
+        def _mock_send_mail_to_admin(email_subject, email_body):
+            """Mocks email_manager.send_mail_to_admin() as its not possible to
+            send mail with self.testapp_swap, i.e with the URLs defined in
+            main_cron.
+            """
+            email_subjects.append(email_subject)
+            email_bodies.append(email_body)
+
+        send_mail_to_admin_swap = self.swap(
+            email_manager, 'send_mail_to_admin', _mock_send_mail_to_admin)
+
+        with self.testapp_swap, send_mail_to_admin_swap:
+            self.get_html_response('/cron/mail/admin/job_status')
+
+        self.assertEqual(email_subjects, ['MapReduce status report'])
+        self.assertEqual(email_bodies, ['All MapReduce jobs are running fine.'])
+
+        self.logout()
+
+    def test_send_mail_to_admin_on_job_failure(self):
+        self.login(self.ADMIN_EMAIL, is_super_admin=True)
+
+        email_subjects = []
+        email_bodies = []
+
+        def _mock_send_mail_to_admin(email_subject, email_body):
+            """Mocks email_manager.send_mail_to_admin() as its not possible to
+            send mail with self.testapp_swap, i.e with the URLs defined in
+            main_cron.
+            """
+            email_subjects.append(email_subject)
+            email_bodies.append(email_body)
+
+        send_mail_to_admin_swap = self.swap(
+            email_manager, 'send_mail_to_admin', _mock_send_mail_to_admin)
+
+        job_id = SampleMapReduceJobManager.create_new()
+        SampleMapReduceJobManager.enqueue(
+            job_id, taskqueue_services.QUEUE_NAME_DEFAULT)
+        self.assertEqual(
+            self.count_jobs_in_taskqueue(
+                taskqueue_services.QUEUE_NAME_DEFAULT), 1)
+
+        self.process_and_flush_pending_tasks()
+        self.assertEqual(
+            SampleMapReduceJobManager.get_status_code(job_id),
+            jobs.STATUS_CODE_COMPLETED)
+
+        shard_state_model_class = mapreduce_model.ShardState
+
+        recent_job_models = shard_state_model_class.all()
+        for job_model in recent_job_models:
+            job_model.retries += 1
+            job_model.put()
+
+        with self.testapp_swap, send_mail_to_admin_swap:
+            self.get_html_response('/cron/mail/admin/job_status')
+
+        self.assertEqual(email_subjects, ['MapReduce failure alert'])
+        self.assertIn(
+            '5 jobs have failed in the past 25 hours. More information '
+            '(about at most 50 jobs; to see more, please check the logs)',
+            email_bodies[0])
```

Comment:
Why [0]? In general try to check everything (in this case, all of email_bodies).

Reply:
I cannot test the whole email body as the message contain fields like last_update_time. Though len(email_bodies) is 1, its just the contents in that single str is quite big, that's why I used assertIn. I added an extra check for len(email_bodies).